File: sepl/public/js/led_conversion_proposal.js
// frappe.ui.form.on('LED Conversion Proposal', {
// 	refresh(frm) {
		
// 	}
// })
frappe.ui.form.on('LED Conversion Proposal', {
	validate: function(frm, cdt, cdn){ 
        var d = locals[cdt][cdn];
        // d.installed_bf_power_conversion = d.units+d.wattage;
        frm.doc.proposal_detals.forEach(function(d) { 
                d.installed_bf_power_conversion = d.units*d.wattage;
                d.power_before_conversion_at_port = (d.units*d.wattage)*d.usage_factor*0.4;
                d.installed_power_af_conversion = d.sepl_wattage*d.qty;
                d.power_after_conversion_at_port = (d.sepl_wattage*d.qty)*d.usage_factor*0.4;
                d.power_consumption_per_year_before_conversion_in_kwh = ((d.power_before_conversion_at_sea+((d.units*d.wattage)*d.usage_factor*0.4)))*24*365/1000;
                d.power_consumption_per_year_after_conversion_in_kwh = ((d.power_after_conversion_at_sea+((d.sepl_wattage*d.qty)*d.usage_factor*0.4)))*24*365/1000
        });
        main_total_qty(frm, cdt, cdn);
        },
})

// frappe.ui.form.on('LED Conversion Proposal Details', {
//         installed_bf_power_conversion: function(frm, cdt, cdn) {
//                 main_total_qty(frm, cdt, cdn);
//         }
// })

function main_total_qty(frm, cdt, cdn) {
        var d = locals[cdt][cdn];
        var installed_bf_power_conversion = 0;
        frm.doc.proposal_detals.forEach(function(d) { installed_bf_power_conversion += d.installed_bf_power_conversion});
        frm.set_value('installed_bf_power_conversion', installed_bf_power_conversion);

        var power_before_conversion_at_sea = 0;
        frm.doc.proposal_detals.forEach(function(d) { power_before_conversion_at_sea += d.power_before_conversion_at_sea});
        frm.set_value('power_before_conversion_at_sea', power_before_conversion_at_sea);

        var power_before_conversion_at_port = 0;
        frm.doc.proposal_detals.forEach(function(d) { power_before_conversion_at_port += d.power_before_conversion_at_port});
        frm.set_value('power_before_conversion_at_port', power_before_conversion_at_port);

        var installed_power_af_conversion = 0;
        frm.doc.proposal_detals.forEach(function(d) { installed_power_af_conversion += d.installed_power_af_conversion});
        frm.set_value('installed_power_af_conversion', installed_power_af_conversion);

        var power_after_conversion_at_sea = 0;
        frm.doc.proposal_detals.forEach(function(d) { power_after_conversion_at_sea += d.power_after_conversion_at_sea});
        frm.set_value('power_after_conversion_at_sea', power_after_conversion_at_sea);

        var power_after_conversion_at_port = 0;
        frm.doc.proposal_detals.forEach(function(d) { power_after_conversion_at_port += d.power_after_conversion_at_port});
        frm.set_value('power_after_conversion_at_port', power_after_conversion_at_port);

        var power_consumption_per_year_before_conversion_in_kwh = 0;
        frm.doc.proposal_detals.forEach(function(d) { power_consumption_per_year_before_conversion_in_kwh += d.power_consumption_per_year_before_conversion_in_kwh});
        frm.set_value('power_consumption_per_year_before_conversion_in_kwh', power_consumption_per_year_before_conversion_in_kwh);

        var power_consumption_per_year_after_conversion_in_kwh = 0;
        frm.doc.proposal_detals.forEach(function(d) { power_consumption_per_year_after_conversion_in_kwh += d.power_consumption_per_year_after_conversion_in_kwh});
        frm.set_value('power_consumption_per_year_after_conversion_in_kwh', power_consumption_per_year_after_conversion_in_kwh);
        frm.set_value('power_savings', power_consumption_per_year_before_conversion_in_kwh-power_consumption_per_year_after_conversion_in_kwh);
        frm.set_value('power_saving_per_year', power_after_conversion_at_port);
        frm.set_value('fuel_saving_per_year', power_after_conversion_at_port*220/1000000);
        frm.set_value('fuel_cost_saving__per_year', frm.doc.fuel_saving_per_year*650);
        frm.set_value('co2_saving_per_year', frm.doc.fuel_saving_per_year*3.5);
        }	

File: sepl/sepl/report/led_conversion_proposal/led_conversion_proposal.py
# Copyright (c) 2024, SEPL and contributors
# For license information, please see license.txt
import frappe
from frappe import _

def execute(filters=None):
    columns, data = get_columns(), get_data(filters)
    return columns, data

def get_columns():
    columns = [
        {"fieldname": "customer", "label": _("Customer"), "fieldtype": "Data", "width": 80},
        {"fieldname": "date", "label": _("Date"), "fieldtype": "Date", "width": 80},
        {"fieldname": "page_no", "label": _("Page No"), "fieldtype": "Data", "width": 80},
        {"fieldname": "sepl_proposal", "label": _("SEPL Proposal"), "fieldtype": "Data", "width": 120},
        {"fieldname": "power_before_conversion_at_port", "label": _("Power Before Conversion (Port)"), "fieldtype": "Float", "width": 120},
        {"fieldname": "light_details", "label": _("Light Details"), "fieldtype": "Data", "width": 120},
        {"fieldname": "sepl_wattage", "label": _("SEPL Wattage"), "fieldtype": "Float", "width": 120},
        {"fieldname": "installed_power_af_conversion", "label": _("Installed Power After Conversion"), "fieldtype": "Float", "width": 120},
        {"fieldname": "location", "label": _("Location"), "fieldtype": "Data", "width": 120},
        {"fieldname": "qty", "label": _("Qty"), "fieldtype": "Int", "width": 120},
        {"fieldname": "power_after_conversion_at_sea", "label": _("Power After Conversion (Sea)"), "fieldtype": "Float", "width": 120},
        {"fieldname": "wattage", "label": _("Wattage"), "fieldtype": "Float", "width": 120},
        {"fieldname": "usage_factor", "label": _("Usage Factor"), "fieldtype": "Float", "width": 120},
        {"fieldname": "power_after_conversion_at_port", "label": _("Power After Conversion (Port)"), "fieldtype": "Float", "width": 120},
        {"fieldname": "fixtures", "label": _("Fixtures"), "fieldtype": "Data", "width": 120},
        {"fieldname": "installed_bf_power_conversion", "label": _("Installed Before Conversion Power"), "fieldtype": "Float", "width": 120},
        {"fieldname": "power_consumption_per_year_before_conversion_in_kwh", "label": _("Annual Power Consumption Before (kWh)"), "fieldtype": "Float", "width": 120},
        {"fieldname": "units", "label": _("Units"), "fieldtype": "Int", "width": 120},
        {"fieldname": "power_before_conversion_at_sea", "label": _("Power Before Conversion (Sea)"), "fieldtype": "Float", "width": 120},
        {"fieldname": "power_consumption_per_year_after_conversion_in_kwh", "label": _("Annual Power Consumption After (kWh)"), "fieldtype": "Float", "width": 120},
    ]
    return columns

def get_data(filters):
    conditions, filter_values = get_conditions(filters)
    
    query = """
        SELECT
            led.customer,
            led.customer_name,
            led.date,
            led.name AS transaction_id,
            ledt.page_no,
            ledt.sepl_proposal,
            ledt.power_before_conversion_at_port,
            ledt.light_details,
            ledt.sepl_wattage,
            ledt.installed_power_af_conversion,
            ledt.location,
            ledt.qty,
            ledt.power_after_conversion_at_sea,
            ledt.wattage,
            ledt.usage_factor,
            ledt.power_after_conversion_at_port,
            ledt.fixtures,
            ledt.installed_bf_power_conversion,
            ledt.power_consumption_per_year_before_conversion_in_kwh,
            ledt.units,
            ledt.power_before_conversion_at_sea,
            ledt.power_consumption_per_year_after_conversion_in_kwh
        FROM `tabLED Conversion Proposal` led
        JOIN `tabLED Conversion Proposal Details` ledt ON led.name = ledt.parent
        WHERE 1=1 {conditions}
        ORDER BY led.date, led.customer ASC
    """.format(conditions=conditions)

    return frappe.db.sql(query, filter_values, as_dict=True)

def get_conditions(filters):
    conditions = []
    filter_values = {}

    if filters.get("customer"):
        conditions.append("AND led.customer = %(customer)s")
        filter_values["customer"] = filters["customer"]

    if filters.get("transaction_id"):
        conditions.append("AND led.name = %(transaction_id)s")
        filter_values["transaction_id"] = filters["transaction_id"]

    if filters.get("from_date"):
        conditions.append("AND led.date >= %(from_date)s")
        filter_values["from_date"] = filters["from_date"]

    if filters.get("to_date"):
        conditions.append("AND led.date <= %(to_date)s")
        filter_values["to_date"] = filters["to_date"]

    return " ".join(conditions), filter_values
